private/bufpkg/bufmodule/bufmoduleref/bufmoduleref.go

```diff
@@ -364,6 +371,46 @@ func ValidateModulePinsUniqueByIdentity(modulePins []ModulePin) error {
 	return nil
 }
 
+// ValidateModulePinsConsistentDigests verifies that module pins to the same commit don't change digests.
+// This is important to avoid MITM issues, where the module digest stored in a buf.lock file doesn't match
+// the module pin returned from the BSR.
+func ValidateModulePinsConsistentDigests(
+	ctx context.Context,
+	bucket storage.ReadWriteBucket,
+	modulePins []ModulePin,
+) error {
+	currentConfig, err := buflock.ReadConfig(ctx, bucket)
+	if err != nil {
+		if storage.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
+	if len(currentConfig.Dependencies) == 0 {
+		return nil
+	}
+	currentIdentityAndCommitToDigest := make(map[string]string, len(currentConfig.Dependencies))
+	for _, dep := range currentConfig.Dependencies {
+		key := fmt.Sprintf("%s/%s/%s:%s", dep.Remote, dep.Owner, dep.Repository, dep.Commit)
+		currentIdentityAndCommitToDigest[key] = dep.Digest
+	}
```

Comment:
A `ModulePin` is also a `ModuleIdentity`. `dep.IdentityString()` would simplify the map key.

Reply:
Yep - I thought it would make it easier to read w/ the symmetry between two different types, but this works.